airflow/operators/python.py

```diff
@@ -251,27 +253,37 @@ def execute(self, context: Context) -> Any:
             self.log.info("Proceeding with downstream tasks...")
             return condition
 
-        downstream_tasks = context["task"].get_flat_relatives(upstream=False)
-        self.log.debug("Downstream task IDs %s", downstream_tasks)
+        if not self.downstream_task_ids:
+            self.log.info("No downstream tasks; nothing to do.")
+            return
 
-        if downstream_tasks:
-            dag_run = context["dag_run"]
-            execution_date = dag_run.execution_date
+        dag_run = context["dag_run"]
+        execution_date = dag_run.execution_date
+        if TYPE_CHECKING:
+            assert isinstance(execution_date, DateTime)
```

Comment:
I think you can do `execution_date = context["logical_date"]`, that one should be guaranteed to be a pendulum DateTime.

Reply:
Logical date is typed as `datetime` too it seems.

And worse, it does not exist on DagRunPydantic, so mypy complains about that.  i'm just keeping the insert

Reply:
Ok actually... at least we can use `cast("DateTime", dag_run.execution_date)` which is a little cleaner than the assert i suppose